contracts/Escrow.sol

```diff
@@ -142,124 +145,135 @@ contract Escrow is ChainlinkClient, Initializable, ContextUpgradeable, OwnableUp
-        require(amount <= IERC20(tokenAddress).allowance(from, vault), "UNAUTHORIZED");
-        return true;
-    }
 
-    
-    function verifyOfferIntegrity(address tokenAddress,  uint256 amount) public pure returns(bool)
-    {
-        require(tokenAddress != address(0), "INVALID TOKENADDR");
-        require(amount > 0  , "MUST BE > 0 AMT");
-        return true;
-    }
+        if (store.executorCounterpart.partyAddress == msg.sender) {
+            store.executorCounterpart.cancel = true;
+        }
+        else {
+            store.creatorCounterpart.cancel = true;
+        }
 
-    function offerPerUser(address u) public view returns(uint256[] memory ) {
-        return (_openOffers[u]);
+        if (store.executorCounterpart.cancel == true && store.creatorCounterpart.cancel == true) {
+            _deals[offerId].offerStatus = OfferStatus.OFFER_CANCELLED;
+            removeOfferForUser(store.executorCounterpart.partyAddress, offerId);
+            removeOfferForUser(store.creatorCounterpart.partyAddress, offerId);
+            prepWithdrawal(offerId, responses[offerId].successCount);
+            withdraw(offerId, 10000);
+        }
+
+        emit OfferCancelled(msg.sender, offerId);
+        return true;
     }
 
-    function removeOfferForUser(address u, uint256 offerId ) private returns(bool)
+    function removeOfferForUser(address user, uint256 offerId) private onlyParticipant(offerId) returns (bool)
     {
-        uint256[] memory userOffers = _openOffers[u];
+        uint256[] memory userOffers = _openOffers[user];
 
-        if(_openOffers[u].length == 1){
-            _openOffers[u][0] = 0;
+        if(_openOffers[user].length == 1){
+            _openOffers[user][0] = 0;
             return true;
         }
         for (uint i = 0; i<userOffers.length-1; i++){
             if(userOffers[i] == offerId ){
-                 _openOffers[u][i] = _openOffers[u][userOffers.length-1];
-                 _openOffers[u].pop();
+                 _openOffers[user][i] = _openOffers[user][userOffers.length-1];
+                 _openOffers[user].pop();
                 return true;
             }   
         }
         return false;
     }
 
-    function setTreasury(address _treasury) public onlyGovernor {
-        require(_treasury != address(0), "INVALID ADDRESS");
-        treasury = ITreasury(_treasury);
-    }
-
     // function that saves time of proof sending
     function saveProof(bytes calldata _proof, uint256 offerId, uint256 targetWindow) public onlyExecutor(offerId) {
         require(_proof.length > 0, "No proof provided"); // check if proof is empty
         require(_deals[offerId].offerStatus == OfferStatus.OFFER_CREATED, "ERROR: OFFER_NOT_ACTIVE");
+        require(block.number < _deals[offerId].dealStartBlock + _deals[offerId].dealLengthInBlocks && block.number > _deals[offerId].dealStartBlock, "Out of block timerange");
 
         _proofblocks[offerId][targetWindow] = block.number;
```

Comment:
I believe the logic that needs to be implemented here is that the window num needs to be calculated from the target window (target window is really a target block). The value that should be stored in ether memory is: 
`_proofblocks[offerId][window_num] = block.number`
where window num is like 0,1, 2, ... deal length whereas target_window is like some block 756411 or whatever

Reply:
Its important especially since the API is just going to iterate from 0 ... num_windows in the deal, and getting the block number stored in memory for each proof

Reply:
indeed, `window_num` != `target_window` - currently reworking.

Reply:
By the way, the window block.number will be stored directly to memory, while the proofWindow count will come from the chainlink api (fulfill)